### ISSUE TITLE
Home Assistant Energy Config

### ISSUE THREAD

**Issue by tuanha2000vn** [StephanJoubert/home_assistant_solarman#1]
Thank for the AWESOME custom component... been looking for a while and that exactly what I need.

Do you use the Energy configuration in Home Assistant? What is the entity for Grid consumption, Return to grid, Battery system, Solar production?

![image](https://user-images.githubusercontent.com/2765517/148453980-74ecca75-47d1-44ee-ada7-2e075a1b1f9c.png)

The Solarman component add 50 entities for me
```
type: entities
entities:
  - entity: sensor.deye_ac_temperature
  - entity: sensor.deye_alert
  - entity: sensor.deye_battery_current
  - entity: sensor.deye_battery_power
  - entity: sensor.deye_battery_soc
  - entity: sensor.deye_battery_status
  - entity: sensor.deye_battery_temperature
  - entity: sensor.deye_battery_voltage
  - entity: sensor.deye_communication_board_version_no
  - entity: sensor.deye_control_board_version_no
  - entity: sensor.deye_current_l1
  - entity: sensor.deye_current_l2
  - entity: sensor.deye_daily_energy_bought
  - entity: sensor.deye_daily_energy_sold
  - entity: sensor.deye_daily_load_consumption
  - entity: sensor.deye_daily_production
  - entity: sensor.deye_dc_temperature
  - entity: sensor.deye_external_ct_l1_power
  - entity: sensor.deye_external_ct_l2_power
  - entity: sensor.deye_gen_connected_status
  - entity: sensor.deye_gen_power
  - entity: sensor.deye_grid_connected_status
  - entity: sensor.deye_grid_status
  - entity: sensor.deye_grid_voltage_l1
  - entity: sensor.deye_grid_voltage_l2
  - entity: sensor.deye_internal_ct_l1_power
  - entity: sensor.deye_internal_ct_l2_power
  - entity: sensor.deye_inverter_id
  - entity: sensor.deye_inverter_l1_power
  - entity: sensor.deye_inverter_l2_power
  - entity: sensor.deye_load_l1_power
  - entity: sensor.deye_load_l2_power
  - entity: sensor.deye_load_voltage
  - entity: sensor.deye_micro_inverter_power
  - entity: sensor.deye_pv1_current
  - entity: sensor.deye_pv1_power
  - entity: sensor.deye_pv1_voltage
  - entity: sensor.deye_pv2_current
  - entity: sensor.deye_pv2_power
  - entity: sensor.deye_pv2_voltage
  - entity: sensor.deye_running_status
  - entity: sensor.deye_smartload_enable_status
  - entity: sensor.deye_time_of_use
  - entity: sensor.deye_total_energy_bought
  - entity: sensor.deye_total_energy_sold
  - entity: sensor.deye_total_grid_power
  - entity: sensor.deye_total_grid_production
  - entity: sensor.deye_total_load_consumption
  - entity: sensor.deye_total_load_power
  - entity: sensor.deye_total_power
  - entity: sensor.deye_total_production
  - entity: sensor.deye_work_mode
title: SolarMAN
```

**Comment by StephanJoubert** [comment]
Thank you for the kind words. 

I don't have much experience with the Energy portion of HA, but the way you have configured your system looks correct to me.

**Comment by tuanha2000vn** [comment]
![image](https://user-images.githubusercontent.com/2765517/148784165-0800683a-7394-4868-ab18-8b909cfb1ae7.png)

Thank again, the custom component provide all detail sensor, after a few tweak, I've managed to make this awesome energy monitor <3

**Comment by StephanJoubert** [comment]
Very impressive. I have been trying with the energy monitor with a little success. Would you mind sharing some pointers how you got the very nice looking dashboard.

**Comment by tuanha2000vn** [comment]
![image](https://user-images.githubusercontent.com/2765517/149378422-533ab8b0-6a52-4586-afbd-64ded2a1133a.png)

Most of the work have been done by your custom component.
Grid consumption: DEYE Daily Energy Bought
Return to grid: DEYE Daily Energy Sold
Solar production: DEYE Daily Production

I've only add 2 custom sensor for battery in and out.

Battery data provide in sensor.deye_battery_power, negative number mean charging, positive number mean discharging. So I've create 2 sensor template for number of W charging/discharing like this:
```
- platform: template
    batt_in:
      friendly_name: "Batt In W"
      value_template: >
        {%- if states.sensor.deye_battery_power.state | int <= 0 -%}
        {{ 0 - states.sensor.deye_battery_power.state | int }}
        {%- else -%}
        0{%- endif %}
      unit_of_measurement: W
      device_class: power
      icon_template: mdi:battery-plus
    batt_out:
      friendly_name: "Batt Out W"
      value_template: >
        {%- if states.sensor.deye_battery_power.state | int > 0 -%}
        {{ states.sensor.deye_battery_power.state }}
        {%- else -%}
        0{%- endif %}
      unit_of_measurement: W
      device_class: power
      icon_template: mdi:battery-minus
```

Use platform integration to convert W to kWh
```
- platform: integration
  source: sensor.batt_in
  name: batt_in_integration
  unit_prefix: k
  round: 3

- platform: integration
  source: sensor.batt_out
  name: batt_out_integration
  unit_prefix: k
  round: 3
```
Then use unility_meter to reset it to zero every day
```
utility_meter:
  batt_in_integration_daily:
    source: sensor.batt_in_integration
    cycle: daily
  batt_out_integration_daily:
    source: sensor.batt_out_integration
    cycle: daily
```
Finally Configuration > Energy > Add Battery system
Energy going into the batterry (kWh): sensor.batt_in_integration_daily
Energy coming out of the batterry (kWh): sensor.batt_out_integration_daily

**Comment by timon65za** [comment]
> ![image](https://user-images.githubusercontent.com/2765517/149378422-533ab8b0-6a52-4586-afbd-64ded2a1133a.png)
> 
> Most of the work have been done by your custom component. Grid consumption: DEYE Daily Energy Bought Return to grid: DEYE Daily Energy Sold Solar production: DEYE Daily Production
> 
> I've only add 2 custom sensor for battery in and out.
> 
> Battery data provide in sensor.deye_battery_power, negative number mean charging, positive number mean discharging. So I've create 2 sensor template for number of W charging/discharing like this:
> 
> ```
> - platform: template
>     batt_in:
>       friendly_name: "Batt In W"
>       value_template: >
>         {%- if states.sensor.deye_battery_power.state | int <= 0 -%}
>         {{ 0 - states.sensor.deye_battery_power.state | int }}
>         {%- else -%}
>         0{%- endif %}
>       unit_of_measurement: W
>       device_class: power
>       icon_template: mdi:battery-plus
>     batt_out:
>       friendly_name: "Batt Out W"
>       value_template: >
>         {%- if states.sensor.deye_battery_power.state | int > 0 -%}
>         {{ states.sensor.deye_battery_power.state }}
>         {%- else -%}
>         0{%- endif %}
>       unit_of_measurement: W
>       device_class: power
>       icon_template: mdi:battery-minus
> ```
> 
> Use platform integration to convert W to kWh
> 
> ```
> - platform: integration
>   source: sensor.batt_in
>   name: batt_in_integration
>   unit_prefix: k
>   round: 3
> 
> - platform: integration
>   source: sensor.batt_out
>   name: batt_out_integration
>   unit_prefix: k
>   round: 3
> ```
> 
> Then use unility_meter to reset it to zero every day
> 
> ```
> utility_meter:
>   batt_in_integration_daily:
>     source: sensor.batt_in_integration
>     cycle: daily
>   batt_out_integration_daily:
>     source: sensor.batt_out_integration
>     cycle: daily
> ```
> 
> Finally Configuration > Energy > Add Battery system Energy going into the batterry (kWh): sensor.batt_in_integration_daily Energy coming out of the batterry (kWh): sensor.batt_out_integration_daily
> 
> ![image](https://user-images.githubusercontent.com/2765517/149380152-7604524c-d959-4b22-9442-c22308590690.png) I've also add Configuration > Intergration > Add Integration Forecast.Solar for a nice prediction graph which display in the Energy section like this

Thanks for the details but can you please help in where you add the scripting.
I have tried putting it in configuration.yaml but I am getting a lot of errors.
Sorry for the dum questions but I am new to HA and it is a new installations.
I did had the energy working until HA upgraded to 2022.2.3 now it is complaining about "Grid neutrality could not be calculated"
Hopefuuly with youre work around i can get it fixed

**Comment by StephanJoubert** [comment]
@timon65za The scripts (templates) goes in configuration.yaml, but I also had problems. It seems that HA changed some of the specifications with the updates and one need to make provision for that. I think you can start with the scripts as provided by @tuanha2000vn , and then google the help for "templates" to see if you can fix the errors. 

I plan to add scripts and blueprints in the repo that people can just use, but need to test it first to see if it works. One blueprint will enable my swimming pool pump only while I am not working off the battery. Almost like a non-critical load, but bit more intelligent.

**Comment by kolakidd** [comment]
I think I solved the above question on placement and formatting:

I made to the code was the addition of sensors: in the platform template. that got that part working. You just have to watch the formatting when cutting and pasting.

Placement is as follows:

On it's own, as in at the top of configuration.yaml or somewhere where it's not inside sensors, binary_sensors etc:

```
utility_meter:
  batt_in_integration_daily:
    source: sensor.batt_in_integration
    cycle: daily
  batt_out_integration_daily:
    source: sensor.batt_out_integration
  cycle: daily
```

 
Then, under sensor:


```
  - platform: template
    sensors:
      batt_in:
        friendly_name: "Batt In W"
        value_template: >
          {%- if states.sensor.deye_battery_power.state | int <= 0 -%}
          {{ 0 - states.sensor.deye_battery_power.state | int }}
          {%- else -%}
          0{%- endif %}
        unit_of_measurement: W
        device_class: power
        icon_template: mdi:battery-plus
      batt_out:
        friendly_name: "Batt Out W"
        value_template: >
          {%- if states.sensor.deye_battery_power.state | int > 0 -%}
          {{ states.sensor.deye_battery_power.state }}
          {%- else -%}
          0{%- endif %}
        unit_of_measurement: W
        device_class: power
        icon_template: mdi:battery-minus
  - platform: integration
    source: sensor.batt_in
    name: batt_in_integration
    unit_prefix: k
    round: 3
  - platform: integration
    source: sensor.batt_out
    name: batt_out_integration
    unit_prefix: k
    round: 3
```


@StephanJoubert thank you for this amazing/lekka work on the integration.

**Comment by havoc2021** [comment]
Hi @StephanJoubert
Thank you for this AWESOME work, I finally got my Deye inverter in HA and all data displaying correctly been looking for weeks for something like this.

@tuanha2000vn  thank you for the only missing piece I needed, one problem and I hope you can help me with it. I could get the battery display in Energy dashboard, but seems like nothing is flowing to the battery but it's currently charging.
![image](https://user-images.githubusercontent.com/87378757/155478431-a88e8f2e-2a34-4e88-abc6-26621af04092.png)

**Comment by kolakidd** [comment]
It took a while for the data to start logging, I think the energy dashboard takes around two hours to populate when you first set it up. The way I checked if it was working was to set up an entity card with battery in and battery out, then click on either of those and you will see a small graph that will show if they are logging power

Entities:
![image](https://user-images.githubusercontent.com/33936947/155484723-3c96158c-3279-433c-af1e-c6d6353be003.png)

What you should see when you click on the entity:
![image](https://user-images.githubusercontent.com/33936947/155484786-61a27395-5a46-44e4-810f-6536e5438d61.png)

**Comment by havoc2021** [comment]
Hi @kolakidd 

You correct after like 30min or so it started showing up :-D

![image](https://user-images.githubusercontent.com/87378757/155485452-2031d405-0e07-451a-af76-27aa2aff37be.png)

THANK YOU ALL!!!

**Comment by StephanJoubert** [comment]
@kolakidd . @timon65za , @tuanha2000vn Just to let you know that I dug around the (very scarce) documentation and found the modbus registers for the battery charge and discharge from the inverter. It is therefore not needed to do the template calculation anymore. My expectation is that this would be a much more  accurate method if there are large battery charge/discharge cycles. It forms part of release 1.3.0 so you can just update using HACS.

**Comment by kolakidd** [comment]
@StephanJoubert awesome! Updated now, will give it a try.

**Comment by havoc2021** [comment]
@StephanJoubert hope you dont mind, I have given it a go as well :-D

**Comment by havoc2021** [comment]
Hi @StephanJoubert 
Thank you for the charge and discharge sensor, I take it this replace the IN/OUT template.
But I see this is TOTAL charge and discharge, anyway you can make it daily?

**Comment by StephanJoubert** [comment]
@havoc2021 , yes this takes the place of the template. Unfortunately this is the only values provided by the inverter, but the  state_class "total_increasing" tells the home assistant dashboard to mark the value at midnight and calculate the daily values. I use on my system, and the dashboard  and graphs are correct. I have updated the documention with the new entities:
See: https://github.com/StephanJoubert/home_assistant_solarman/blob/main/energy.md

If one needs a daily value for another reason than the dashboard, it could be done with the "utility_meter" component from what I can see - have not done that myself.

**Comment by kolakidd** [comment]
Just as a note, I updated and then tracked the four values (Modbus source VS Template) in Graphana. The template values are very good, but not 100% accurate. For instance there is a slight overlap between charge and discharge, which shouldn't be possible, because a battery can't simultaneously charge and discharge. This is not to say that they won't give you a good idea.

I've updated to the Modbus values on the dashboard this morning.
